pkg/operator/controllers/azurensg/subnetnsg.go

```diff
@@ -0,0 +1,139 @@
+package azurensg
+
+// Copyright (c) Microsoft Corporation.
+// Licensed under the Apache License 2.0.
+
+import (
+	"context"
+	"encoding/json"
+	"fmt"
+	"strings"
+
+	mgmtnetwork "github.com/Azure/azure-sdk-for-go/services/network/mgmt/2019-07-01/network"
+	"github.com/Azure/go-autorest/autorest/azure"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/runtime"
+	azureproviderv1beta1 "sigs.k8s.io/cluster-api-provider-azure/pkg/apis/azureprovider/v1beta1"
+
+	"github.com/Azure/ARO-RP/pkg/api"
+	arov1alpha1 "github.com/Azure/ARO-RP/pkg/operator/apis/aro.openshift.io/v1alpha1"
+	"github.com/Azure/ARO-RP/pkg/util/aad"
+	"github.com/Azure/ARO-RP/pkg/util/azureclient/mgmt/network"
+	"github.com/Azure/ARO-RP/pkg/util/clusterauthorizer"
+	"github.com/Azure/ARO-RP/pkg/util/subnet"
+)
+
+type subnetDescriptor struct {
+	resourceGroup string
+	vnetName      string
+	subnetName    string
+}
+
+func (r *AzureNSGReconciler) reconcileSubnetNSG(ctx context.Context, instance *arov1alpha1.Cluster) error {
+	// Get endpoints from operator
+	azEnv, err := azure.EnvironmentFromName(instance.Spec.AZEnvironment)
+	if err != nil {
+		return err
+	}
+	// Grab azure-credentials from secret
+	credentials, err := clusterauthorizer.AzCredentials(ctx, r.kubernetescli)
+	if err != nil {
+		return err
+	}
+	resource, err := azure.ParseResourceID(instance.Spec.ResourceID)
+	if err != nil {
+		return err
+	}
+	// create service principal token from azure-credentials
+	token, err := aad.GetToken(ctx, r.log, string(credentials.ClientID), string(credentials.ClientSecret), string(credentials.TenantID), azEnv.ActiveDirectoryEndpoint, azEnv.ResourceManagerEndpoint)
+	if err != nil {
+		return err
+	}
+	// create refreshable authorizer from token
+	authorizer, err := clusterauthorizer.NewAuthorizer(token)
+	if err != nil {
+		return err
+	}
+	subnetsClient := network.NewSubnetsClient(&azEnv, resource.SubscriptionID, authorizer)
+	// the main logic starts here
+	subnets, masterResourceGroup, err := r.getSubnets(ctx)
+	if err != nil {
+		return err
+	}
+	for ws := range subnets {
+		err = r.fixSubnet(ctx, api.ArchitectureVersion(instance.Spec.ArchitectureVersion), resource.SubscriptionID, ws.vnetName, ws.subnetName, ws.resourceGroup, masterResourceGroup, instance.Spec.InfraID, subnets[ws], subnetsClient)
+		if err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+func (r *AzureNSGReconciler) fixSubnet(ctx context.Context, architectureVersion api.ArchitectureVersion, subscriptionID, vnetName, subnetName, vnetResourceGroup, resourcesResourceGroup, infraID string, isWorkerSubnet bool, subnetsClient network.SubnetsClient) error {
+	fullSubnet, err := subnetsClient.Get(ctx, vnetResourceGroup, vnetName, subnetName, "")
```

Comment:
fullSubnet is a bad name, I changed it to subnetObject to differentiate it from the package name. I don't like one letter var names for things other than iterators, is that a golang pattern that we should follow instead?

Reply:
We use one letter names for things that aren't iterators in a couple places in this repo, although it is a little underdescriptive. We use `_env` for `env` when necessary to avoid conflict because it's also the package name.

`subnetObject` is fine.